rdkit/Chem/RegistrationHash.py

```diff
@@ -0,0 +1,502 @@
+    ats = tuple(atRanks[x] for x in sg.GetAtoms())
+    if sortAtomOrder:
+        ats = tuple(sorted(ats))
+    bnds = tuple(bndOrder[x] for x in sg.GetBonds())
+
+    res = dict(fieldName=fieldName, atom=ats, bonds=bnds, value=data)
+    return res
+
+
+def _GetCanononicalBondRep(bond, atomRanks):
+    aid1 = bond.GetBeginAtomIdx()
+    aid2 = bond.GetEndAtomIdx()
+    if atomRanks[aid1] > atomRanks[aid2] or (atomRanks[aid1] == atomRanks[aid2]
+                                             and aid1 > aid2):
+        aid1, aid2 = aid2, aid1
+    return (aid1, aid2)
+
+
+def _CanonicalizeSRUSGroup(mol, sg, atRanks, bndOrder, sortAtomAndBondOrder):
+    """
+    NOTES: if sortAtomAndBondOrder is true then the atom and bond lists will be sorted.
+    This assumes that the ordering of those lists is not important
+
+    """
+    if sg.GetProp("TYPE") != "SRU":
+        return None
+    ats = tuple(atRanks[x] for x in sg.GetAtoms())
+    bnds = tuple(bndOrder[x] for x in sg.GetBonds())
+    if sortAtomAndBondOrder:
+        ats = tuple(sorted(ats))
+        bnds = tuple(sorted(bnds))
+    props = sg.GetPropsAsDict()
+    res = dict(
+        type="SRU",
+        atoms=ats,
+        bonds=bnds,
+        index=props.get("index", 0),
+        connect=props.get("CONNECT", ""),
+        label=props.get("LABEL", ""),
+    )
+    if "PARENT" in props:
+        res["PARENT"] = props["PARENT"]
+    if "XBHEAD" in props:
+        xbhbonds = tuple(
+            _GetCanononicalBondRep(mol.GetBondWithIdx(x), atRanks)
+            for x in props["XBHEAD"])
+        if sortAtomAndBondOrder:
+            xbhbonds = tuple(sorted(xbhbonds))
+        res["XBHEAD"] = xbhbonds
+    if "XBCORR" in props:
+        xbcorrbonds = tuple(
+            _GetCanononicalBondRep(mol.GetBondWithIdx(x), atRanks)
+            for x in props["XBCORR"])
+        if len(xbcorrbonds) % 2:
+            raise ValueError("XBCORR should have 2N bonds")
+        if sortAtomAndBondOrder:
+            # these are pairs, so we need to sort them as such
+            tmp = []
+            for i in range(0, len(xbcorrbonds), 2):
+                b1, b2 = xbcorrbonds[i], xbcorrbonds[i + 1]
+                if b1 > b2:
+                    b1, b2 = b2, b1
+                tmp.append((b1, b2))
+            xbcorrbonds = tuple(sorted(tmp))
+        res["XBCORR"] = xbcorrbonds
+    return res
+
+
+def _CaonicalizeCOPSGroup(sg, atRanks, sortAtomAndBondOrder):
+    """
+    NOTES: if sortAtomAndBondOrder is true then the atom and bond lists will be sorted.
+    This assumes that the ordering of those lists is not important
+    """
+    if sg.GetProp("TYPE") != "COP":
+        return None
+    ats = tuple(atRanks[x] for x in sg.GetAtoms())
+    if sortAtomAndBondOrder:
+        ats = tuple(sorted(ats))
+    props = sg.GetPropsAsDict()
+    # we are intentionally not doing the label here. Marvin adds one, biovia draw does not
+    res = dict(type="COP", atoms=ats, index=props.get("index", 0))
```

Comment:
We aren't doing bonds here, which is probably ok, but maybe we should at least generate a warning if the SGroup does have bonds set?

Reply:
I updated to do bonds